File: 1472-design-browser-history/1472-design-browser-history.py
class BrowserHistory:
    class Page :
        def __init__(self,url):
            self.url = url
            self.next = None
            self.prev = None 

    def __init__(self, homepage: str):
        self.home = BrowserHistory.Page(homepage)
        self.cur = self.home 


        

    def visit(self, url: str) -> None:
        nextPage = BrowserHistory.Page(url)
        if self.cur :
            self.cur.next = nextPage 
            nextPage.prev = self.cur
            self.cur = self.cur.next
        else :
            self.cur = nextPage
        return self.cur.url
        

    def back(self, steps: int) -> str:
        while self.cur.prev and steps > 0 :
            self.cur = self.cur.prev 
            steps -= 1
        return self.cur.url
        

    def forward(self, steps: int) -> str:
        while self.cur.next and steps > 0:
            self.cur = self.cur.next 
            steps -= 1
        return self.cur.url

        


# Your BrowserHistory object will be instantiated and called as such:
# obj = BrowserHistory(homepage)
# obj.visit(url)
# param_2 = obj.back(steps)
# param_3 = obj.forward(steps)

File: 1472-design-browser-history/1472-design-browser-history.java
class BrowserHistory {
    Page cur ;
    class Page{
        String url ;
        Page next , prev ;
        public  Page(String url){
            this.url = url ;
            this.prev = null ;
            this.next = null ;

        }

    }
    
    

    public BrowserHistory(String homepage) {
        cur = new Page(homepage) ;

        
    }
    
    public void visit(String url) {
        Page nextPage = new Page(url) ;
        nextPage.prev = cur ;
        cur.next = nextPage ;
        cur = cur.next ;

        
    }
    
    public String back(int steps) {
        while(cur.prev != null && steps > 0) {
            steps--;
            cur = cur.prev ;
        }
        return cur.url ;
        
    }
    
    public String forward(int steps) {
        while(cur.next != null && steps > 0) {
            steps--;
            cur = cur.next ;
        }
        return cur.url ;
        
    }
}

/**
 * Your BrowserHistory object will be instantiated and called as such:
 * BrowserHistory obj = new BrowserHistory(homepage);
 * obj.visit(url);
 * String param_2 = obj.back(steps);
 * String param_3 = obj.forward(steps);
 */
